input_common: Redesign mouse panning

bm01 opened the pull request:
I wanted to give a try at fixing the issues raised by #10420 (whose author is away for some months and can't update the PR), and I ended up with a somewhat more conservative solution.

Attempts to fix the following issues:

1. Sensitivity setting affecting the stick's range, making either the max value unreachable or min value too high if not set to 50%.
Fixed by applying the sensitivity on the mouse input alone, before any clamping.

2. Very small movements being impossible to perform.
Fixed by removing the line that sets the mouse event's length to a minimum of 3. Testing shows that it is rarely higher than 1 even with a low mouse polling rate, so setting it to 3 meant some details were lost.

3. Stick "resistance" feeling inconsistent.
This is more subjective. The buffer decay is applied in both `Move` (9% per mouse event) and `UpdateStickInput` (4% per update tick). Comment says "Average mouse movements" but this is really just a double exponential decay that depends not just on the buffer's current value and how much time has passed (for example I think a high polling rate mouse meant a stronger decay).
Fixed by applying it in `UpdateStickInput` only.

4. Slow stick recentering time causing slippery (high end inertia) movements.
`mouse_panning_timeout` was used to recenter a stick after a certain amount of update with no user input, but a high value resulted in slippery movements and a low one tended to make the stick recenter too easily during slow movements (killing any build-up). This can be fixed by changing the decay function and make the recenter time from a high length must faster without affecting the time it takes to recenter from a short length too much. The following graph shows the decay (Y) applied to `last_mouse_change` on each update based on the stick length (X). (Update: was slightly tweaked to be less aggressive)
![chart](https://github.com/yuzu-emu/yuzu/assets/1279721/873ed67a-8396-4cc9-8bc7-1ce7b1a7c1fa)

5. Lack of configurability.
Only the sensitivity could be configured, and globally.
The stick's decay strength and minimum amount can now be configured to fit the user preference. Another setting was added to help counteract against a game's built-in deadzone. This last setting and sensitivity were split for both X and Y axis. Because the user may want to have different settings between games, the configure button was moved from the advanced widget to the first player input widget, and values are saved in the profiles.

6. https://github.com/yuzu-emu/yuzu/issues/10481
Implemented what was suggested. The enable setting is no longer saved but is still read (effectively a read-only setting).

This still isn't perfect by any mean:
- Qt's mouse event appears to lack granularity, but I don't think we have much choice.
- The update interval is somewhat slow (10ms) so buffer decay is never going to feel very smooth.
- Buffer decay is tied to the update interval so it can't be changed without making it either slower or faster.
- Probably as a consequence of above, slippery movement currently cannot be completely prevented without introducing various issues.

german77 commented:
Oh nice. I will take a look soon. From a quick glance did you removed the initial deadzone? 

This part in specific https://github.com/yuzu-emu/yuzu/blob/53fd08b360a393af2e6bfdddc5c549492fcb40bf/src/input_common/drivers/mouse.cpp#L145-L150

bm01 commented:
> From a quick glance did you removed the initial deadzone?

I did. It caused the second issue I mention in the PR. In my personal experience most of the mouse movements triggered that deadzone because the length is almost always inferior to 1, but this could be because I play with 1800 DPI mouse and 1000 Hz polling rate. I tried lowering them and I didn't feel like removing it caused any undesirable effect. Is it really needed?

german77 commented:
I haven't tested yet. But I will provide a few games where the lack of deadzone is an issue just give me a few hours.

german77 commented:
Had a chance to test. Here you can see SMO doesn't react to the mouse at all even when it's moving quite a bit. The sensitivity seems a bit low. I can't quite reach the borders unless I really move fast the mouse.
https://github.com/yuzu-emu/yuzu/assets/5944268/4be213bb-d876-4429-aff2-1a67dc47d8f8

german77 commented:
Much better this time. I will try to poke the code a bit to make the stick go to the center even with a deadzone but it's overall good. 

UI Config sucks. We probably need to make a separate category for mouse config since those two input boxes have no explanation XD.

bm01 commented:
I tried to figure a proper solution to that but it was either recentering the stick after an arbitrarily high amount of time, or take the risk of doing it too early and make small movements harder to perform, due to clearing any buffer build-up too low to actually result in any movement but actually contributing towards the next (if that makes any sense).

I agree about the UI. I'll give a try at improving it. How about a "Configure" button that opens a new window similar to "Motion / Touch"? I was also thinking about separating horizontal and vertical sensitivity and deadzone (assuming I can figure out the maths), and also exposing the two float here that now control stick decay:
https://github.com/yuzu-emu/yuzu/blob/465d7c830c165a47d03d3f3a8dd196bc694c4ba5/src/input_common/drivers/mouse.cpp#L112-L113

bm01 commented:
Here's how it could look:

![Screenshot 2023-06-01 064002](https://github.com/yuzu-emu/yuzu/assets/1279721/76d7132c-384b-44e4-9df9-3e1131f0ef64)

Interestingly, users can choose to set a decay strength of 0% and a minimum decay of 5%, and effectively get something that's closer to what the original formula was.

With split sensitivity settings I'm not sure what should be done about this though:
https://github.com/yuzu-emu/yuzu/blob/9e7656a597d4a1942a4db3ae996e321766a69d76/src/input_common/drivers/mouse.cpp#L122-L126

bm01 commented:
![Screenshot 2023-06-01 170319](https://github.com/yuzu-emu/yuzu/assets/1279721/719c0a1f-6286-4305-9240-fc13eb63a7af)

Does it expect `2.0f` here? I didn't get any warning about it in Visual Studio. Anyway I don't want to trigger too many actions :p

german77 commented:
Yes. 2.0f. Don't worry about the actions. they are there to help you make sure nothing breaks.

bm01 commented:
I just noticed the deadzone setting in Controls > Player is affecting mouse panning too, which isn't necessarily a problem but for some reason I ended up with one and my tests were done thinking there was none.

Setting it to 0% allows for even finer controls in TOTK, which appears to have a lower built-in deadzone than whatever I had.

Just a thought but maybe a warning could be added in that new widget if the deadzone isn't set to of 0 and range to 100, because I don't think there's any reason to have them to any other value when mouse panning is enabled. But I couldn't find a way to get those values from `ConfigureMousePanning`.

bm01 commented:
Moved the configure button from the advanced widget to the first player input widget, and made it so that the settings are saved in the profiles.

Added a warning that appears only the right stick deadzone and range aren't set to the default values.

![Screenshot 2023-06-03 203226](https://github.com/yuzu-emu/yuzu/assets/1279721/48fc464c-3454-41d0-a813-ef62a056e621)

Updated the PR's original post to reflect those changes since it didn't mention the new settings at all.

bunnei commented:
@german77 ping to take a look at the recent changes here!

german77 reviewed (approved):
Sorry for the delay. It's working great now.
